### PR TITLE
WIP - d3 data viz experiments

### PR DESCRIPTION
**Pull request by owocki**
spent a few cycles last night building a data tool that visualizes where the work is on the platform

slack-imgs

screen shot 2018-04-13 at 7 39 01 pm
screen shot 2018-04-13 at 7 38 45 pm

we could maybe use this for content marketing purposes

screen recording 2018-04-13 at 08 24 pm

### CODE REVIEW

**Comment by mkosowsk** [comment]
Looks great! And nice use of D3, another jewel of a library in the wonderful JavaScript ecosystem 😍😍😍

File: app/marketing/views.py [hunk 1]
```diff
@@ -19,12 +19,14 @@
 from __future__ import unicode_literals
 
 import json
+import math
+import random
 
 from django.contrib.admin.views.decorators import staff_member_required
```
Comment (owocki) [inline app/marketing/views.py]: donno if theres an easy way to programmatically cleanu pthe imports...

**Comment by owocki** [comment]
@mbeacom can i get an 👀 on this sometime today?

File: app/dataviz/models.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+# Create your models here.
+from economy.models import SuperModel
+
+
+class DataPayload(SuperModel):
+    """Define the structure of an ExternalBounty."""
+
+    key = models.CharField(db_index=True, max_length=255, help_text=_("key for this data report"))
+    report = models.CharField(max_length=255, blank=True, help_text=_("The report associated with this project"))
+    payload = JSONField(default=False, blank=True)
```
Comment (mbeacom) [inline app/dataviz/models.py]: Can we default this to an empty dict?

File: app/dataviz/templates/dataviz/circles.html [hunk 3]
```diff
@@ -0,0 +1,153 @@
+{% extends "admin/base_site.html" %}
```
Comment (mbeacom) [inline app/dataviz/templates/dataviz/circles.html]: We'll want to strip some of the whitespace and additional lines throughout the new template files.

File: app/dataviz/views.py [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+from django.contrib.admin.views.decorators import staff_member_required
+from django.template.response import TemplateResponse
+from django.utils import timezone
+
+from chartit import Chart, DataPool
+from dashboard.models import Profile, UserAction
+from marketing.models import EmailEvent, EmailSubscriber, GithubEvent, SlackPresence, SlackUser, Stat
+
+
+
```
Comment (mbeacom) [inline app/dataviz/views.py]: Too many new lines

File: app/marketing/models.py [hunk 5]
```diff
@@ -64,6 +64,20 @@ class Meta:
     def __str__(self):
         return f"{self.key}: {self.val}"
 
+    @property
+    def val_since_yesterday(self):
+        try:
+            return self.val - Stat.objects.filter(key=self.key, created_on__lt=self.created_on, created_on__hour=self.created_on.hour).order_by('-created_on').first().val
+        except:
```
Comment (mbeacom) [inline app/marketing/models.py]: Bare exceptions

File: app/retail/templates/admin/index.html [hunk 6]
```diff
@@ -71,16 +71,19 @@
...
+            <a href="{% url 'stats' %}">KPI Dashboard</a>
         </ul>
         <ul>
             <a href="{% url 'cohort' %}">Cohort Analysis</a>
         </ul>
         <ul>
             <a href="{% url 'funnel' %}">Funnel Analysis</a>
         </ul>
+        <ul>
+            <a href="/_administration/viz/">Analytics Playground</a>
```
Comment (mbeacom) [inline app/retail/templates/admin/index.html]: When possible, we should use `url()` in the event the url path changes.

File: app/dataviz/views.py [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+            # tables
+            params['tables'][t] = source
+
+            # charts
+            # compute avg
+            total = 0
+            count = source.count() - 1
+            avg = "NA"
+            if count > 1:
+                for i in range(0, count):
```
Comment (mbeacom) [inline app/dataviz/views.py]: We can probably use `enumerate()` for some of the range iterations.

File: app/dataviz/migrations/0003_datapayload_payload.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
+# Generated by Django 2.0.3 on 2018-04-14 23:53
```
[the remark this answers is not in the document]
Comment (owocki) [reply app/dataviz/migrations/0003_datapayload_payload.py]: b95b7bec0bc37d878a09f8548a54ef230b44a939

File: app/dataviz/tests.py [hunk 9]
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
```
[the remark this answers is not in the document]
Comment (owocki) [reply app/dataviz/tests.py]: 2e7c6de9419d9fd34af5a32b17e2ab1b0041c686

File: app/dataviz/templates/dataviz/circles.html (hunk 3 above)
Comment (owocki) [reply app/dataviz/templates/dataviz/circles.html]: there a programatic way to do it?  `make fix` isnt seeming to do the trick

File: app/marketing/models.py (hunk 5 above)
Comment (owocki) [reply app/marketing/models.py]: ?

File: app/dataviz/views.py (hunk 4 above)
Comment (owocki) [reply app/dataviz/views.py]: updated

File: app/retail/templates/admin/index.html (hunk 6 above)
Comment (owocki) [reply app/retail/templates/admin/index.html]: updated

File: app/dataviz/views.py (hunk 7 above)
Comment (owocki) [reply app/dataviz/views.py]: ncie.. i didnt know about enumerate

File: app/dataviz/models.py (hunk 2 above)
Comment (owocki) [reply app/dataviz/models.py]: udpated

File: app/dataviz/views.py [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+            'period_size': period_size,
+        }
+    }
+    return TemplateResponse(request, 'cohort.html', params)
+
+
+def funnel_helper_get_data(key, k, daily_source, weekly_source, start_date, end_date):
+    if key == 'sessions':
+        return sum(daily_source.filter(key='google_analytics_sessions_gitcoin', created_on__gte=start_date, created_on__lt=end_date).values_list('val', flat=True))
+    if key == 'email_subscribers':
```
[the remark this answers is not in the document]
Comment (owocki) [reply app/dataviz/views.py]: i kind of like having this explicitly laid out.. do you feel strongly about this?

File: app/marketing/views.py (hunk 1 above)
Comment (owocki) [reply app/marketing/views.py]: thanks!

**Comment by owocki** [comment]
just responded to your code review @mbeacom

**Pull request merged**